README


This archive contains the distribution Module-Path,
version 0.13:

  get the full path to a locally installed module

This software is copyright (c) 2013 by Neil Bowers.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.013.



Changes
Revision history for Perl module Module::Path

0.10_01 2014-02-16
    - Testsuite now uses Cwd::abs_path() on paths from %INC,
      to cope with synlinks in @INC directories.
      This was causing test failures on Debian(-based) systems.

0.10 2014-02-04
    - mpath can display paths for multiple modules (Ahmad Syaltut)
    - specified min perl version 5.6.0

0.09_01 2013-08-21
    - If a directory in @INC is a symlink, return the linked-to directory
      in the path. Problem report and patch from Sharl Morlaroll
      https://github.com/neilbowers/Module-Path/issues/4

0.09 2013-01-20
    - You can now pass partial module paths, in the format
      used as the keys in %INC. Suggested by Steven Haryanto.

0.08 2012-10-18
    - t/03-mpath.t: now used Devel::FindPerl to get path to perl,
      instead of $^X. Thanks to LeonT.
    - should be quoting the path to the mpath binary in t/03-mpath.t
      Thanks to Graham Knop (haarg).

0.07 2012-09-24
    - Made the repository path a link in the pod (Olivier Mengué)
    - Updated pod to refer to review of this and similar modules.
      Nudged by Olivier Mengué.

0.06 2012-09-18
    - should check $? after running backticks in t/03-mpath.t,
      to ensure process exited cleanly.

0.05 2012-09-18
    - Added lightweight mpath script, which just calls module_path()
      Made path to script portable, thanks to help from kentnl and leont.
      Ugh, test could be run with a different perl than mpath.
      More help, from leont, ether and haarg.

0.04 2012-08-30
    - github repo was specified wrongly in dist.int (Olivier Mengué)

0.03 2012-08-27
    - Directory separator should be / and not \ on MSWin32.
      Thanks to kmx for helping me identify the issue.

0.02 2012-08-26
    - bugfix in initialisation of separator (Tom Molesworth)
    - Added github repo to metadata and documentation
    - SEE ALSO now covers all modules I'm aware of that provide
      the same functionality

0.01 2012-08-20
    - initial release of module



lib/Module/Path.pm
package Module::Path;
# ABSTRACT: get the full path to a locally installed module
$Module::Path::VERSION = '0.12';
use 5.006;
use strict;
use warnings;
use File::Basename 'dirname';
use Cwd qw/ abs_path /;

require Exporter;

our @ISA       = qw(Exporter);
our @EXPORT_OK = qw(module_path);

my $SEPARATOR;

BEGIN {
    if ($^O =~ /^(dos|os2)/i) {
        $SEPARATOR = '\\';
    } elsif ($^O =~ /^MacOS/i) {
        $SEPARATOR = ':';
    } else {
        $SEPARATOR = '/';
    }
}

sub module_path
{
    my $module = shift;
    my $relpath;
    my $fullpath;

    ($relpath = $module) =~ s/::/$SEPARATOR/g;
    $relpath .= '.pm' unless $relpath =~ m!\.pm$!;

    DIRECTORY:
    foreach my $dir (@INC) {
        next DIRECTORY if not defined($dir);

        # see 'perldoc -f require' on why you might find
        # a reference in @INC
        next DIRECTORY if ref($dir);

        # The directory path might have a symlink somewhere in it,
        # so we get an absolute path (ie resolve any symlinks).
        # The previous attempt at this only dealt with the case
        # where the final directory in the path was a symlink,
        # now we're trying to deal with symlinks anywhere in the path.
        $dir = abs_path($dir);
        next DIRECTORY unless defined($dir);

        $fullpath = $dir.$SEPARATOR.$relpath;
        return $fullpath if -f $fullpath;
    }

    return undef;
}

1;

=head1 NAME

Module::Path - get the full path to a locally installed module

=head1 SYNOPSIS

 use Module::Path 'module_path';
 
 $path = module_path('Test::More');
 if (defined($path)) {
   print "Test::More found at $path\n";
 } else {
   print "Danger Will Robinson!\n";
 }

=head1 DESCRIPTION

Module::Path provides a single function, C<module_path()>,
which will find where a module is installed locally.

It works by looking in all the directories in C<@INC>
for an appropriately named file:

=over 4

=item

Foo::Bar becomes C<Foo/Bar.pm>, using the correct directory path
separator for your operating system.

=item

Iterate over C<@INC>, ignoring any references
(see L<"perlfunc"/"require"> if you're surprised to hear
that you might find references in C<@INC>).

=item

For each directory in C<@INC>, append the partial path (C<Foo/Bar.pm>),
again using the correct directory path separator.
If the resulting file exists, return this path.

=item

If a directory in C<@INC> is a symlink, then we resolve the path,
and return a path containing the linked-to directory.

=item

If no file was found, return C<undef>.

=back

I wrote this module because I couldn't find an alternative
which dealt with the points listed above, and didn't pull in
what seemed like too many dependencies to me.

The distribution for C<Module::Path> includes the C<mpath>
script, which lets you get the path for a module from the command-line:

 % mpath Module::Path

The C<module_path()> function will also cope if the module name includes C<.pm>;
this means you can pass a partial path, such as used as the keys in C<%INC>:

  module_path('Test/More.pm') eq $INC{'Test/More.pm'}

The above is the basis for one of the tests.

=head1 BUGS

Obviously this only works where the module you're after has its own C<.pm>
file. If a file defines multiple packages, this won't work.

This also won't find any modules that are being loaded in some special
way, for example using a code reference in C<@INC>, as described
in L<"perlfunc"/"require">.


=head1 SEE ALSO

There are a number of other modules on CPAN which provide the
same or similar functionality:
L<App::whichpm>,
L<Class::Inspector>,
L<Module::Data>,
L<Module::Filename>,
L<Module::Finder>,
L<Module::Info>,
L<Module::Locate>,
L<Module::Mapper>,
L<Module::Metadata>,
L<Module::Runtime>,
L<Module::Util>,
and L<Path::ScanINC>.

I've written a review of all such modules that I'm aware of:

=over 4

L<http://neilb.org/reviews/module-path.html>

=back

Module::Path was written to be fast, portable, and have a low number of
core-only runtime dependencies. It you only want to look up the path to
a module, it's a good choice.

If you want more information, such as the module's version, what functions
are provided, etc, then start by looking at L<Module::Info>,
L<Module::Metadata>, and L<Class::Inspector>.

The following scripts can also give you the path:
L<perldoc>,
L<whichpm|https://www.metacpan.org/module/whichpm>.


=head1 REPOSITORY

L<https://github.com/neilbowers/Module-Path>

=head1 AUTHOR

Neil Bowers E<lt>neilb@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Neil Bowers <neilb@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.



Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.013.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "get the full path to a locally installed module",
  "AUTHOR" => "Neil Bowers <neil\@bowers.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Module-Path",
  "EXE_FILES" => [
    "bin/mpath"
  ],
  "LICENSE" => "perl",
  "NAME" => "Module::Path",
  "PREREQ_PM" => {
    "Cwd" => 0,
    "Exporter" => 0,
    "File::Basename" => 0,
    "Getopt::Long" => 0,
    "Pod::Usage" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Devel::FindPerl" => 0,
    "File::Spec::Functions" => 0,
    "FindBin" => "0.05",
    "Test::More" => "0.88"
  },
  "VERSION" => "0.13",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Cwd" => 0,
  "Devel::FindPerl" => 0,
  "Exporter" => 0,
  "File::Basename" => 0,
  "File::Spec::Functions" => 0,
  "FindBin" => "0.05",
  "Getopt::Long" => 0,
  "Pod::Usage" => 0,
  "Test::More" => "0.88",
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



